contracts/margin/external/exchangewrappers/ZeroExV2ExchangeWrapper.sol

```diff
@@ -0,0 +1,295 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+*/
+
+pragma solidity 0.4.24;
+pragma experimental "v0.5.0";
+pragma experimental ABIEncoderV2;
+
+import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import { IExchange } from "../../../external/0x/v2/protocol/Exchange/interfaces/IExchange.sol";
+import { LibFillResults } from "../../../external/0x/v2/protocol/Exchange/libs/LibFillResults.sol";
+import { LibOrder } from "../../../external/0x/v2/protocol/Exchange/libs/LibOrder.sol";
+import { MathHelpers } from "../../../lib/MathHelpers.sol";
+import { TokenInteract } from "../../../lib/TokenInteract.sol";
+import { ExchangeReader } from "../../interfaces/ExchangeReader.sol";
+import { ExchangeWrapper } from "../../interfaces/ExchangeWrapper.sol";
+
+
+/**
+ * @title ZeroExV2ExchangeWrapper
+ * @author dYdX
+ *
+ * dYdX ExchangeWrapper to interface with 0x Version 2
+ */
+contract ZeroExV2ExchangeWrapper is
+    LibFillResults,
+    LibOrder,
+    ExchangeWrapper,
+    ExchangeReader
+{
+    using SafeMath for uint256;
+    using TokenInteract for address;
+
+    // ============ State Variables ============
+
+    // msg.senders that will put the correct tradeOriginator in callerData when doing an exchange
+    mapping (address => bool) public TRUSTED_MSG_SENDER;
+
+    // address of the ZeroEx V1 Exchange
+    address public ZERO_EX_EXCHANGE;
+
+    // address of the ZeroEx V1 TokenTransferProxy
+    address public ZERO_EX_TOKEN_PROXY;
+
+    // address of the ZRX token
+    address public ZRX;
+
+    // ============ Constructor ============
+
+    constructor(
+        address zeroExExchange,
+        address zeroExProxy,
+        address zrxToken,
+        address[] trustedMsgSenders
+    )
+        public
+    {
+        ZERO_EX_EXCHANGE = zeroExExchange;
+        ZERO_EX_TOKEN_PROXY = zeroExProxy;
+        ZRX = zrxToken;
+
+        for (uint i = 0; i < trustedMsgSenders.length; i++) {
+            TRUSTED_MSG_SENDER[trustedMsgSenders[i]] = true;
+        }
+
+        // The ZRX token does not decrement allowance if set to MAX_UINT
+        // therefore setting it once to the maximum amount is sufficient
+        // NOTE: this is *not* standard behavior for an ERC20, so do not rely on it for other tokens
+        ZRX.approve(ZERO_EX_TOKEN_PROXY, MathHelpers.maxUint256());
+    }
+
+    // ============ Public Functions ============
+
+    function exchange(
+        address tradeOriginator,
+        address receiver,
+        address makerToken,
+        address takerToken,
+        uint256 requestedFillAmount,
+        bytes orderData
+    )
+        external
```

Comment:
I'm thinking about if this needs to be `nonReentrant` I don't think it does, but I'd like to be more convinced. I also see the V1 exchange wrapper is also not `nonReentrant`